design/036-package_id.md

```diff
@@ -0,0 +1,107 @@
+
+Beyond the importance of having different binaries for different situations, and not "overwriting" binaries that should have different ``package_id``s, it also became evident that a more accurate modeling for computing "what needs to be rebuilt and what not" was necessary. Because the use cases for Conan have been growing in size, and there are many teams with large dependency graphs, and doing CI at scale with such graphs can be challenging, and become inefficient if not modeled correctly. A better ``package_id``, together with easier to use ``lockfiles`` was necessary for enabling such CI at scale.
+
+Other feedback along 1.X versions also showed that for some teams it is also important to take into account the information from ``build_requires`` into the consumers ``package_ids``, because some tools new versions could indeed produce different binaries. Factoring in the influence of ``build_requires`` in Conan 1.X was impossible, a chicken-and-egg problem, because ``build_requires`` in 1.X are evaluated after the ``package_ids`` have been computed. In Conan 2.0, the dependency graph is always fully expanded, including ``build_requires``, and after that the ``package_ids`` are computed. That enabled the possibility of fulfilling that need.
+
+
+## Detailed Design
+
+Conan 2.0 will implement a new ``package_id`` default strategy which is dynamic, and can take different values for the versions of the dependencies, depending on the context and usage. The three main cases are:
+
+- "embed" case: this mode will be used whenever some binary is "inlining" or "embedding" native binary code or resources of its dependency in its own artifacts. This happens when a shared library links a static library, when an executable links a static library, or when a translation unit of a library includes a header file from another library and inlines it. This mode typically means that when the dependency changes, the consumer has to be rebuilt to embed the changed dependency in its binary.
+- "non embed" case: this mode happens when the resources and artifacts from the dependency are not inlined nor embedded in the current artifacts. An executable using a shared library or a static library "linking" (it actually doesn’t link, but this terminology can be used in build systems, to express a link time dependency) another static library. This mode means that in many situations when the interfaces of the dependency don't change, the binaries of the consumer don't need to be rebuilt when the binaries of the dependency change. If the interfaces of the dependency change, then bumping its version patch, minor, major, can define to its consumers if they need to rebuild (bumping the minor) or not (bumping the patch), or if they should expect breaking changes (bumping the major).
```

Comment:
> static library "linking" 

I tried to express this in my first comment, but was afraid of using word "linking". Maybe it's a good idea to give some vocabulary / definition in the beginning. What is "linking" and different kind of dependency nesting: we talk at least about 3 levels now - the package in question (for which package_id is computed) and it's up- and down-stream dependencies. Maybe it's a good idea to define "package" (for example as binary files produced by building the Conan package) or just "binaries".